torch/optim/optimizerw.py

```diff
@@ -0,0 +1,95 @@
+from .optimizer import Optimizer
+from .adam import Adam
+from .sgd import SGD
+
+
+class OptimizerW(Optimizer):
+    r"""Base class for decoupled weight decay optimizers.
+
+    The distinction between weight decay and L2 regularisation is discussed in
+    `Decoupled Weight Decay Regularization`_.
+
+    .. warning::
+        Parameters need to be specified as collections that have a deterministic
+        ordering that is consistent between runs. Examples of objects that don't
+        satisfy those properties are sets and iterators over values of dictionaries.
+
+    Arguments:
+        params (iterable): an iterable of :class:`torch.Tensor` s or
+            :class:`dict` s. Specifies what Tensors should be optimized.
+        defaults: (dict): a dict containing default values of optimization
+            options (used when a parameter group doesn't specify them).
+
+    .. _Decoupled Weight Decay Regularization:
+        https://arxiv.org/abs/1711.05101
+    """
+
+    def step(self, closure=None):
+        """Performs a single optimization step (parameter update).
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss. Optional for most optimizers.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            lr = group['lr']
+            _group = dict(group)
+            weight_decay = _group.pop('weight_decay')
+            for p in group['params']:
+                if p.grad is None:
+                    continue
+
+                if p.grad.is_sparse:
+                    update = self.get_sparse_update(p, **_group)
+                else:
+                    update = self.get_update(p, **_group)
+
+                p.data.mul_(1 - weight_decay * lr).add_(-lr, update)
```

Comment:
The main point of `OptimizerW` is that it does not do weight decay through L2 regularisation (as it is implemented in most optimisers at this point), but instead decouples the weight decay from the gradient.

A more interesting line for universal weight decay (in the spirit of backward compatibility) would look like
```
weight_decay = _group.pop('weight_decay', 0.)
if weight_decay > 0:
    grad = grad.add(weight_decay, p.data)
```

Reply:
Indeed, this is a more interesting way of phrasing the weight_decay.

Implementing L2 regularization in a way similar to the weight_decay would be nice and allow for the Optimizer and OptimizerW to be combined? The way the weight decay is exposed to the user though would still require separate Optimizers. Anyway, this doesn't need to be part of the PR.

Reply:
Not sure what to do with this one. It would of course be possible to move the L2 regularization in `Optimizer.step`, but then `p.grads` seems to be everywhere. I would generally consider it cleaner if it were possible to pass only the grads to `get_update` (since `get_update` generally does not need the current parameter values). However, since `self.state` is indexed by the parameters, there seems to be no other way.

After all, ideal semantics for `get_update` would be to convert the raw gradients to the actual update. Any thoughts on semantics and/or interface of `get_update`?